src/main/java/org/opensearch/knn/index/query/ResultUtil.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.knn.index.query;
+
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.ScoreDoc;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.search.TotalHits;
+import org.apache.lucene.util.BitSet;
+import org.apache.lucene.util.DocIdSetBuilder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.PriorityQueue;
+
+/**
+ * Utility class used for processing results
+ */
+public final class ResultUtil {
+
+    public static final ResultUtil DEFAULT = new ResultUtil();
+
+    /**
+     * Reduce the results to only include the top k results across all leaf results
+     *
+     * @param perLeafResults Results from the list
+     * @param k the number of results across all leaf results to return
+     * @return Reduced results across all leaf results
+     */
+    public List<Map<Integer, Float>> reduceToTopK(List<Map<Integer, Float>> perLeafResults, int k) {
+        // Iterate over all scores to get min competitive score
+        PriorityQueue<Float> topKScoresQueue = new PriorityQueue<>(k);
+        for (int i = 0; i < k; i++) {
+            topKScoresQueue.add(-Float.MAX_VALUE);
+        }
+
+        for (Map<Integer, Float> perLeafResult : perLeafResults) {
+            for (Float score : perLeafResult.values()) {
```

Comment:
This seems expensive, is there a possibility that we can maintain a global priority queue of with a pair of id in a  and score and keep adding the results in `firstPassSearchLeaf` when we are looping through for filtering deleted docs? At the end of parallel execution we should be able to just poll the priority queue and rescore. 

Its a high level thought and in a way inspired from Lucene's `TopKnnCollectorManager`. There is a slight modification though where we need doc ids (with leaf index I suppose). Worth time boxing it as we can reduce latencies significantly when concurrent search is on

Reply:
@navneet1v had similar idea. I think this is probably a good optimization to make, but will take in a future PR.